### DISCUSSION TITLE
Overloading Keymaps (Switch Choice/Dynamic Node External Update)

### DISCUSSION THREAD

**Discussion by evesdropper** [L3MON4D3/LuaSnip#885] [Q&A] [answered]
I was thinking about using the same mapping for changing choices and dynamic node external update. Is there a way to create a mapping that does so - somewhat like nvim-cmp's super-tab functionality?

I was thinking something like:
```lua
-- psuedocode
vim.keymap.set("i", "<key>", function ()
if active choicenode then
<Plug>luasnip-next-choice
else
dynamic node external update
end
end
```
Is this possible, and if so, how would I make it work?

**Comment by L3MON4D3** [comment, marked as answer]
I think you pretty much have it, use `ls.choice_active()`, `ls.change_choice(1)` and `_G.dynamic_node_external_update(i)` instead of the placeholders, and it should work

**Reply by evesdropper** [reply]
Thanks! That worked perfectly. Let's hope I can say the same for the snippet :pray:

**Reply by L3MON4D3** [reply]
:eyes:

**Reply by evesdropper** [reply]
:sob:  it does not work unfortunately. Perhaps I will go ~~beg for help~~ continue this thread:

Getting the following errors - I think the latter is related to external dynamic node, but it always appears after the first error.
![err](https://github.com/L3MON4D3/LuaSnip/assets/82856360/d590d870-8610-47ca-9bea-96cc2a74a093)
![err2](https://github.com/L3MON4D3/LuaSnip/assets/82856360/44eead04-dc6f-4ed6-817f-1451e531a79f)

Snippet Source:
```lua
local generate_matchcase_dynamic = function(args, snip)
    if not snip.num_cases then
        snip.num_cases = 0
    end
    local cap_cases = tonumber(snip.captures[1])
    snip.num_cases = snip.num_cases + cap_cases
    local nodes = {}
    local ins_idx = 1
    for j = 1, snip.num_cases do
        vim.list_extend(nodes, fmta([[
        case <>:
            <>
        ]], {r(ins_idx, "var" .. j, i(1)), r(ins_idx+1, "pass", i(0))}))
        table.insert(nodes, t({"", ""}))
        ins_idx = ins_idx + 2
    end
    table.remove(nodes, #nodes)
    return isn(nil, nodes, "\t")
end

    s({ trig='mc(%d+)', name='match case', dscr='match n cases', regTrig = true, hidden = true},
    fmta([[
    match <>:
        <>
        <>
    ]],
    { i(1), d(2, generate_matchcase_dynamic, {}, {
	user_args = {
		function(snip) snip.num_cases = snip.num_cases + 1 end,
		function(snip) snip.num_cases = math.max(snip.num_cases - 1, 1) end
	}
    }),
    c(3, {t(""), isn(nil, fmta([[
    case _:
        <>
    ]], {i(1, "pass")}), "\t")})}
    )),
```

**Reply by L3MON4D3** [reply]
Oh, one error I can spot is that in c(3), the isn needs a jump-index (it can only be nil if it is immediately inside the choices-list). Does `isn(1, ...)` fix both issues?
BTW interesting idea to return `isn` from dynamicNode, haven't seen that until now👌

**Reply by evesdropper** [reply]
> Oh, one error I can spot is that in c(3), the isn needs a jump-index (it can only be nil if it is immediately inside the choices-list). Does isn(1, ...) fix both issues?

Hmm, doesn't seem to - still getting the same error :frowning: 

> BTW interesting idea to return isn from dynamicNode, haven't seen that until nowok_hand

Kind of out of necessity as the indents would not persist :sob:

**Reply by L3MON4D3** [reply]
Oh wait, my bad, that `isn` is immediately inside the choices-list, I mixed up the indent :sweat_smile:
But I've found it, it's because the `"pass"`-key is duplicated for `restoreNode`, which is not allowed. Appending `.. j`, like for `"var"` will do it.

**Reply by evesdropper** [reply]
Ah, got it, thanks so much for the help! I also noticed when that works that there's now a small bug with the snippet, which I've just fixed. Now regex captures and external update can exist in harmony :partying_face:

**Reply by L3MON4D3** [reply]
Beatiful!

> Ah, got it, thanks so much for the help!

Of course :D
